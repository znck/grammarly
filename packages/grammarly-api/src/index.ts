export * from './auth'

export * from './checks/check'
export * from './checks/checkGrammar'
export * from './checks/checkPlagiarism'

export * from './transport/enums/AlertFeedbackType'
export * from './transport/enums/AlertImpactType'
export * from './transport/enums/AlertViewType'
export * from './transport/enums/AsyncChecksTypes'
export * from './transport/enums/AutoCorrectFeedbackType'
export * from './transport/enums/AutocompleteFeedbackType'
export * from './transport/enums/DialectType'
export * from './transport/enums/DocumentAudienceType'
export * from './transport/enums/DocumentDomainType'
export * from './transport/enums/DocumentGoalType'
export * from './transport/enums/EmotionFeedbackType'
export * from './transport/enums/ErrorCodeType'
export * from './transport/enums/ErrorSeverityType'
export * from './transport/enums/FeatureType'
export * from './transport/enums/LensFeedbackType'
export * from './transport/enums/MutedFeedbackType'
export * from './transport/enums/OptionType'
export * from './transport/enums/PredictionType'
export * from './transport/enums/SuggestionRejectionReasonType'
export * from './transport/enums/SynonymFeedbackType'
export * from './transport/enums/SystemFeedbackType'
export * from './transport/enums/TakeawayFeedbackType'
export * from './transport/enums/UserMutedScopeType'
export * from './transport/enums/WritingEmotionType'
export * from './transport/enums/WritingStyleType'
export * from './transport/enums/WritingToneType'

export * from './transport/events/AlertEvent'
export * from './transport/events/AlertsChangedEvent'
export * from './transport/events/AsyncCheckFinishedEvent'
export * from './transport/events/CompleteEvent'
export * from './transport/events/EmotionsEvent'
export * from './transport/events/ErrorEvent'
export * from './transport/events/FinishedEvent'
export * from './transport/events/HeatmapEvent'
export * from './transport/events/PlagiarismEvent'
export * from './transport/events/RemoveEvent'
export * from './transport/events/TakeawaysEvent'
export * from './transport/events/TextInfoEvent'
export * from './transport/events/TextMapsEvent'

export * from './transport/interfaces/AlertCardLayout'
export * from './transport/interfaces/AlertExtraProperties'
export * from './transport/interfaces/DocumentContext'
export * from './transport/interfaces/DocumentStatistics'
export * from './transport/interfaces/Emotion'
export * from './transport/interfaces/FeedbackType'
export * from './transport/interfaces/FluencyExtraProperties'
export * from './transport/interfaces/HeatmapRange'
export * from './transport/interfaces/IdAlert'
export * from './transport/interfaces/IdHeatmap'
export * from './transport/interfaces/IdRevision'
export * from './transport/interfaces/IdTakeaway'
export * from './transport/interfaces/OutcomeScores'
export * from './transport/interfaces/OutcomeScoresWithPlagiarism'
export * from './transport/interfaces/PlagiarismExtraProperties'
export * from './transport/interfaces/Synonym'
export * from './transport/interfaces/SynonymsGroup'
export * from './transport/interfaces/VoxExtraProperties'

export * from './transport/messages/AlertFeedbackRequest'
export * from './transport/messages/AlertFeedbackResponse'
export * from './transport/messages/DebugInfoRequest'
export * from './transport/messages/DebugInfoResponse'
export * from './transport/messages/EmotionFeedbackRequest'
export * from './transport/messages/EmotionFeedbackResponse'
export * from './transport/messages/LensFeedbackRequest'
export * from './transport/messages/LensFeedbackResponse'
export * from './transport/messages/MutedFeedbackRequest'
export * from './transport/messages/MutedFeedbackResponse'
export * from './transport/messages/OptionRequest'
export * from './transport/messages/OptionResponse'
export * from './transport/messages/PingRequest'
export * from './transport/messages/PingResponse'
export * from './transport/messages/SetContextRequest'
export * from './transport/messages/SetContextResponse'
export * from './transport/messages/StartRequest'
export * from './transport/messages/StartResponse'
export * from './transport/messages/SubmitOTChunkRequest'
export * from './transport/messages/SubmitOTChunkResponse'
export * from './transport/messages/SubmitOTRequest'
export * from './transport/messages/SubmitOTResponse'
export * from './transport/messages/SynonymsRequest'
export * from './transport/messages/SynonymsResponse'
export * from './transport/messages/SystemFeedbackRequest'
export * from './transport/messages/SystemFeedbackResponse'
export * from './transport/messages/TextStatsRequest'
export * from './transport/messages/TextStatsResponse'
export * from './transport/messages/ToggleChecksRequest'
export * from './transport/messages/ToggleChecksResponse'

export * from './transport/ot/ChangeSet'
export * from './transport/ot/Delta'
export * from './transport/ot/Op'
export * from './transport/ot/OpDelete'
export * from './transport/ot/OpInsert'
export * from './transport/ot/OpRetain'
export * from './transport/ot/Range'
export * from './transport/ot/Transform'

export * from './transport/Request'
export * from './transport/RequestKind'
export * from './transport/Response'
export * from './transport/ResponseKind'

export * from './GrammarlyClient'
export * from './SocketClient'
export * from './SocketError'
