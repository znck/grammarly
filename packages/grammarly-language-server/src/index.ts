import 'reflect-metadata'
import 'isomorphic-fetch'
import { init } from '@grammarly/sdk'
import { Container } from 'inversify'
import { createConnection, ProposedFeatures } from 'vscode-languageserver/node'
import { CLIENT, CLIENT_INFO, CONNECTION, GRAMMARLY_SDK, SERVER } from './constants'
import { CodeActionService } from './services/CodeActionService'
import { ConfigurationService } from './services/ConfigurationService'
import { DiagnosticsService } from './services/DiagnosticsService'
import { DocumentService } from './services/DocumentService'
import { HoverService } from './services/HoverService'

interface Disposable {
  dispose(): void
}

export function startLanguageServer(): void {
  const disposables: Disposable[] = []
  const capabilities: any = {}
  const container = new Container({
    autoBindInjectable: true,
    defaultScope: 'Singleton',
  })
  const connection = createConnection(ProposedFeatures.all)

  container.bind(CONNECTION).toConstantValue(connection)
  container.bind(SERVER).toConstantValue(capabilities)

  connection.onInitialize(async (params) => {
    const options = params.initializationOptions as { clientId: string } | undefined
    if (options?.clientId == null) throw new Error('clientId is required')
    const sdk = await init(options.clientId)

    function onHandleOAuthUrl(url: string): void {
      connection.sendNotification('openOAuthUrl', url)
    }

    container.bind(CLIENT).toConstantValue(params.capabilities)
    container.bind(CLIENT_INFO).toConstantValue({ ...params.clientInfo, id: options.clientId })
    container.bind(GRAMMARLY_SDK).toConstantValue(sdk)

    disposables.push(
      container.get(ConfigurationService).register(),
      container.get(DocumentService).register(),
      container.get(DiagnosticsService).register(),
      container.get(HoverService).register(),
      container.get(CodeActionService).register(),
    )

    return {
      serverInfo: {
        name: 'Grammarly',
      },
      capabilities,
    }
  })

  connection.onExit(() => {
    disposables.forEach((disposable) => disposable.dispose())
  })

  connection.listen()
}
