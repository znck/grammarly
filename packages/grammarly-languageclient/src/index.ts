import type { BaseLanguageClient, MessageTransports } from 'vscode-languageclient'
import type { GrammarlyLanguageClientOptions } from './GrammarlyLanguageClientOptions'
import type { Protocol } from './protocol'
export type { GrammarlyLanguageClientOptions } from './GrammarlyLanguageClientOptions'
export declare class GrammarlyLanguageClient extends BaseLanguageClient {
  protected getLocale(): string
  protected createMessageTransports(encoding: string): Promise<MessageTransports>
  constructor(serverPath: string, options: GrammarlyLanguageClientOptions)
  public readonly protocol: Protocol
}
