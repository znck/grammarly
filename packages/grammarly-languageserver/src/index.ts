export declare function startLanguageServer(): void
