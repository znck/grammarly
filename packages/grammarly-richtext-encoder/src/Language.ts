import { InsertOperation, RichText, RichTextAttributes } from '@grammarly/sdk'
import Parser from 'web-tree-sitter'

export type SourceMap = Array<[original: number, generated: number, length: number]>
export interface Transformer {
  encode(tree: Parser.Tree): [RichText, SourceMap]
  decode(text: RichText): string
}

export function createTransformer(options: {
  isBlockNode(node: Parser.SyntaxNode): boolean
  shouldIgnoreSubtree(node: Parser.SyntaxNode): boolean
  getAttributesFor(node: Parser.SyntaxNode, parentAttrs: RichTextAttributes): RichTextAttributes
  processNode(node: Parser.SyntaxNode, insert: (text: string, node?: Parser.SyntaxNode, skip?: number) => void): void
  stringify(node: Decoder.Node, content: string): string
}): Transformer {
  return { encode, decode }

  function encode(tree: Parser.Tree): [RichText, SourceMap] {
    let offset = 0
    let attributes: RichTextAttributes = {}
    const richtext: RichText = { ops: [] }
    const sourcemap: SourceMap = []
    processNode(tree.rootNode)
    return [richtext, sourcemap]

    function processNode(node: Parser.SyntaxNode): void {
      if (options.shouldIgnoreSubtree(node)) return // stop processing sub-tree
      const previousAttributes = attributes
      attributes = { ...previousAttributes, ...options.getAttributesFor(node, { ...attributes }) }
      options.processNode(node, insert)
      node.children.forEach(processNode)
      if (options.isBlockNode(node) && !hasTrailingNewline()) insert('\n')
      attributes = previousAttributes
    }

    function insert(text: string, node?: Parser.SyntaxNode, skip: number = 0): void {
      if (text === '') return
      richtext.ops.push({
        insert: text,
        attributes: text === '\n' ? pickBlockAttributes(attributes) : pickInlineAttributes(attributes),
      })
      if (node != null) sourcemap.push([node.startIndex + skip, offset, text.length])
      else if (sourcemap.length > 0) {
        const last = sourcemap[sourcemap.length - 1]
        sourcemap.push([last[0] + last[2], offset, 0])
      } else {
        sourcemap.push([0, offset, 0])
      }
      offset += text.length
    }

    function hasTrailingNewline(): boolean {
      return richtext.ops.length > 0 && String(richtext.ops[richtext.ops.length - 1].insert).endsWith('\n')
    }
  }

  function decode(text: RichText): string {
    const ops: InsertOperation[] = text.ops.reduce((ops, op) => {
      if (ops.length === 0 || op.insert === '\n') {
        ops.push(op)
      } else {
        const last = ops[ops.length - 1]
        if (
          typeof last.insert === 'string' &&
          typeof op.insert === 'string' &&
          last.insert !== '\n' &&
          sameAttributes(last.attributes, op.attributes)
        ) {
          last.insert += op.insert
        } else {
          ops.push(op)
        }
      }

      return ops
    }, [] as InsertOperation[])
    const root: Decoder.Element = { type: 'block', childNodes: [], attributes: {}, value: {} }

    function findParent(node: Decoder.Element, attributes: RichTextAttributes): Decoder.Element {
      if (node === root) {
        const parent: Decoder.Element = { type: 'inline', attributes, value: attributes, childNodes: [], parent: node }
        root.childNodes.push(parent)
        return parent
      }

      const diff = diffAttributes(node.attributes, attributes)
      if (diff.removed.length == 0 && diff.added.length === 0) return node
      if (diff.removed.length === 0) {
        const value: RichTextAttributes = {}
        diff.added.forEach((key) => {
          value[key] = attributes[key] as any
        })
        const parent: Decoder.Element = { type: 'inline', attributes, value, childNodes: [], parent: node }
        node.childNodes.push(parent)
        return parent
      }

      if (node.parent == null) throw new Error('Unexpected')

      return findParent(node.parent, attributes)
    }

    let current: Decoder.Element = root
    const leaves: Decoder.Text[] = []
    for (let i = 0; i < ops.length; i += 1) {
      const op = ops[i]
      if (op.insert === '\n' && i > 0) {
        let j = i
        for (; j >= 0; --j) {
          if (ops[j - 1]?.insert === '\n') break
        }
        const target = commonAncestor(leaves.find((leave) => leave.op === ops[j])!, leaves[leaves.length - 1])
        const parent = target.parent!
        const node: Decoder.Element = {
          type: 'block',
          attributes: op.attributes ?? {},
          value: op.attributes ?? {},
          childNodes: [target],
          parent: parent,
        }
        const index = parent.childNodes.indexOf(target)
        parent.childNodes.splice(index, 1)
        parent.childNodes.push(node)
        target.parent = node
      } else {
        const node: Decoder.Text = { type: '#text', op }
        leaves.push(node)
        current = findParent(current, op.attributes ?? {})
        current.childNodes.push(node)
      }
    }

    return processNode(root)

    function processNode(node: Decoder.Node): string {
      if (node.type === '#text') return options.stringify(node, '')
      return options.stringify(node, node.childNodes.map((node) => processNode(node)).join(''))
    }
  }
}

function commonAncestor(a: Decoder.Node, b: Decoder.Node): Decoder.Element {
  const pa = pathFromRoot(a)
  const pb = pathFromRoot(b)
  const n = Math.min(pa.length, pb.length)
  for (let i = 0; i < n; ++i) {
    if (pa[i] !== pb[i]) return getEl(pa[i - 1])
  }

  throw new Error('No commont ancestor')

  function getEl(node: Decoder.Node): Decoder.Element {
    if (node.type === '#text') return node.parent!
    return node
  }
}

function pathFromRoot(node: Decoder.Node): Decoder.Node[] {
  const path: Decoder.Node[] = []
  let current: Decoder.Node | undefined = node
  while (current != null) {
    path.push(current)
    current = current.parent
  }

  return path.reverse()
}

export namespace Decoder {
  interface BaseNode {
    type: string
    parent?: Element
  }

  export interface Text extends BaseNode {
    type: '#text'
    op: InsertOperation
  }

  export interface Element extends BaseNode {
    type: 'block' | 'inline'
    childNodes: Node[]
    attributes: RichTextAttributes
    value: RichTextAttributes
  }

  export type Node = Text | Element
}

function sameAttributes(a?: RichTextAttributes, b?: RichTextAttributes): boolean {
  if (a === b) return true
  if (a == null || b == null) return false
  if (Object.keys(a).length !== Object.keys(b).length) return false

  return (Object.keys(a) as Array<keyof RichTextAttributes>).every((key) => a[key] === b[key])
}

function diffAttributes(
  target?: RichTextAttributes,
  source?: RichTextAttributes,
): { added: Array<keyof RichTextAttributes>; removed: Array<keyof RichTextAttributes> } {
  if (target == null && source == null) return { added: [], removed: [] }
  else if (target == null && source != null) return { added: Object.keys(source) as any, removed: [] }
  else if (target != null && source == null) return { added: [], removed: Object.keys(target) as any }
  else if (target != null && source != null) {
    const added: Array<keyof RichTextAttributes> = []
    const removed: Array<keyof RichTextAttributes> = []

    for (const key of Object.keys(target) as Array<keyof RichTextAttributes>) {
      if (!(key in source)) {
        removed.push(key)
      } else if (source[key] !== target[key]) {
        added.push(key)
        removed.push(key)
      }
    }

    for (const key of Object.keys(source) as Array<keyof RichTextAttributes>) {
      if (!(key in target)) {
        added.push(key)
      }
    }

    return { added, removed }
  }

  return { added: [], removed: [] }
}

function pickBlockAttributes(attributes: RichTextAttributes): RichTextAttributes {
  const picked: RichTextAttributes = {}

  for (const key of ['header', 'list', 'indent'] as const) {
    if (key in attributes) {
      // @ts-expect-error
      picked[key] = attributes[key]
    }
  }

  return picked
}

function pickInlineAttributes(attributes: RichTextAttributes): RichTextAttributes {
  const picked: RichTextAttributes = {}

  for (const key of ['bold', 'italic', 'underline', 'code', 'link'] as const) {
    if (key in attributes) {
      // @ts-expect-error
      picked[key] = attributes[key]
    }
  }

  return picked
}
