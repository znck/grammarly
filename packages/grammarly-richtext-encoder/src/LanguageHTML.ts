import { RichTextAttributes } from '@grammarly/sdk'
import { createTransformer } from './Language'

const IGNORED_NODES = new Set([
  "'",
  '/>',
  '"',
  '<',
  '<!',
  '</',
  '=',
  '>',
  'attribute',
  'attribute_name',
  'attribute_value',
  'comment',
  'doctype',
  'end_tag',
  'erroneous_end_tag',
  'erroneous_end_tag_name',
  'fragment',
  'quoted_attribute_value',
  'raw_text',
  'script_element',
  'self_closing_tag',
  'start_tag',
  'style_element',
  'tag_name',
])
const OTHER_NODES = new Set(['text'])
const BLOCK_NODES = new Set(['element'])

export const html = createTransformer({
  isBlockNode(node) {
    return BLOCK_NODES.has(node.type)
  },
  shouldIgnoreSubtree(node) {
    return IGNORED_NODES.has(node.type)
  },
  getAttributesFor(node) {
    switch (node.type) {
      case 'strong_emphasis':
        return { bold: true }
      case 'emphasis':
        return { italic: true }
      case 'code_span':
        return { code: true }
      case 'atx_heading':
        if (node.firstChild != null) {
          // atx_h[1-6]_marker
          return { header: parseInt(node.firstChild.type.substring(5, 6), 10) as 1 | 2 | 3 | 4 | 5 | 6 }
        }
        return {}

      default:
        return {}
    }
  },
  stringify(node, content) {
    if (node.type === '#text') {
      if (typeof node.op === 'string') return node.op
      return '\n'
    }

    return toHTML(content, node.value)

    function toHTML(text: string, attributes: RichTextAttributes): string {
      if (attributes.bold) return `<b>${text}</b>`
      if (attributes.italic) return `<i>${text}</i>`
      if (attributes.code) return `<code>${text}</code>`
      if (attributes.linebreak) return `<br />`
      if (attributes.link) return `<a href=${JSON.stringify(attributes.link)}>${text}</a>`
      if (attributes.header) return `<h${attributes.header}>${content}</h${attributes.header}>`

      return text
    }
  },
  processNode(node, insert) {
    if (node.type === 'text') {
      insert(node.text, node)
    }
  },
})
