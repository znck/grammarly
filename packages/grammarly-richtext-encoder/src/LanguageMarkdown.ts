import type { RichTextAttributes } from '@grammarly/sdk'
import { createTransformer } from './Language'

const IGNORED_NODES = new Set([
  'atx_h1_marker',
  'atx_h2_marker',
  'atx_h3_marker',
  'atx_h4_marker',
  'atx_h5_marker',
  'atx_h6_marker',
  'block_quote',
  'code_fence_content',
  'fenced_code_block',
  'html_atrribute',
  'html_attribute_key',
  'html_attribute_value',
  'html_cdata_section',
  'html_close_tag',
  'html_comment',
  'html_declaration_name',
  'html_declaration',
  'html_open_tag',
  'html_processing_instruction',
  'html_self_closing_tag',
  'html_tag_name',
  'link_destination',
  'link_reference_definition',
  'list_marker',
  'setext_h1_underline',
  'setext_h2_underline',

  'table_cell',
  'table_column_alignment',
  'table_data_row',
  'table_delimiter_row',
  'table_header_row',
  'table',
  'task_list_item_marker',
])
const OTHER_NODES = new Set([
  'backslash_escape',
  'character_reference',
  'code_span',
  'email_autolink',
  'emphasis',
  'hard_line_break',
  'heading_content',
  'image',
  'info_string',
  'line_break',
  'link_label',
  'link_title',
  'link',
  'link_text',
  'loose_list',
  'soft_line_break',
  'strikethrough',
  'strong_emphasis',
  'text',
  'uri_autolink',
  'virtual_space',
  'www_autolink',
])
const BLOCK_NODES = new Set([
  'document',
  'atx_heading',
  'setext_heading',
  'task_list_item',
  'html_block',
  'image_description',
  'indented_code_block',
  'list_item',
  'paragraph',
  'thematic_break',
  'tight_list',
])

export const markdown = createTransformer({
  isBlockNode(node) {
    return BLOCK_NODES.has(node.type)
  },
  shouldIgnoreSubtree(node) {
    return IGNORED_NODES.has(node.type)
  },
  getAttributesFor(node, attributes) {
    switch (node.type) {
      case 'strong_emphasis':
        return { bold: true }
      case 'emphasis':
        return { italic: true }
      case 'code_span':
        return { code: true }
      case 'atx_heading':
        if (node.firstChild != null) {
          // atx_h[1-6]_marker
          return { header: parseInt(node.firstChild.type.substring(5, 6), 10) as 1 | 2 | 3 | 4 | 5 | 6 }
        }
        return {}
      case 'setext_heading':
        if (node.lastNamedChild != null) {
          // setext_h[1-2]_underline
          return {
            header: parseInt(node.lastNamedChild.type.substring('setext_h'.length, 'setext_h'.length + 1), 10) as 1 | 2,
          }
        }
        return {}
      case 'link':
        if (node.lastNamedChild?.type === 'link_destination') {
          return {
            link: node.lastNamedChild.firstNamedChild?.text ?? '',
          }
        }
        return { link: '' }

      case 'tight_list':
        return {
          list: /[0-9]/.test(node.firstNamedChild?.firstNamedChild?.text ?? '') ? 'number' : 'bullet',
          indent: attributes.indent != null ? attributes.indent + 1 : 1,
        }

      default:
        return {}
    }
  },
  stringify(node, content) {
    if (node.type === '#text') {
      if (typeof node.op.insert === 'string') return node.op.insert
      return '\n'
    }

    return toMarkdown(content, node.value)
  },
  processNode(node, insert) {
    if (node.type === 'text') {
      if (node.parent?.parent?.type === 'atx_heading' && node.parent.firstChild?.equals(node)) {
        insert(node.text.slice(1), node, 1)
      } else if (
        node.parent?.type === 'paragraph' &&
        node.text.startsWith(': ') &&
        node.parent.firstChild?.equals(node)
      ) {
        insert(node.text.slice(2), node, 2) // #255 - Definition Lists
      } else insert(node.text, node)
    } else if (node.type === 'line_break' || node.type === 'hard_line_break') {
      insert('\n', node)
    } else if (node.type === 'soft_line_break') {
      insert(' ', node)
    }
  },
})

function toMarkdown(text: string, attributes?: RichTextAttributes): string {
  if (attributes == null) return text
  if (attributes.bold) return `**${text}**`
  if (attributes.italic) return `_${text}_`
  if (attributes.code) return '`' + text + '`'
  if (attributes.link) return `[${text}](${attributes.link})`
  if (attributes.list && attributes.indent != null)
    return `${'  '.repeat(attributes.indent - 1)}${attributes.list === 'number' ? '1. ' : '- '}${text}\n`
  if (attributes.header) return '#'.repeat(attributes.header) + ' ' + text + '\n'
  if (attributes['code-block']) return '```\n' + text + '\n```\n'
  return text
}
