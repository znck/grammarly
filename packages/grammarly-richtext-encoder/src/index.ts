import Parser from 'web-tree-sitter'
import { html } from './LanguageHTML'
import { markdown } from './LanguageMarkdown'

export type { SourceMap, Transformer } from './Language'

const parsers = new Map<string, Parser>()
const parsersPending = new Map<string, Promise<Parser>>()

export async function createParser(language: string): Promise<Parser> {
  const previous = parsers.get(language)
  if (previous != null) return previous
  const parser = createParserInner()
  parsersPending.set(language, parser)

  return await parser

  async function createParserInner() {
    await Parser.init()

    const parser = new Parser()
    parser.setLanguage(await Parser.Language.load(getLanguageFile()))
    parsers.set(language, parser)
    parsersPending.delete(language)

    return parser
  }

  function getLanguageFile(): string | Uint8Array {
    if (typeof process !== 'undefined' && process.versions?.node != null) {
      if (process.env.NODE_ENV === 'test') {
        return require.resolve(`../dist/tree-sitter-${language}.wasm`)
      }
      return require.resolve(`./tree-sitter-${language}.wasm`)
    }

    return `tree-sitter-${language}.wasm`
  }
}

export const transformers = { html, markdown } as const
