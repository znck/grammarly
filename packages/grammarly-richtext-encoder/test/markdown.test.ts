import { createParser, transformers } from '../src'
import { readFile } from 'node:fs/promises'
import { RichText } from '@grammarly/sdk'

expect.addSnapshotSerializer({
  test(value) {
    return typeof value !== 'string'
  },
  print(richtext, _print, indent) {
    return (richtext as RichText).ops
      .map((op) =>
        indent(
          JSON.stringify(op.insert).slice(1, -1) +
            (op.attributes == null || Object.keys(op.attributes).length === 0
              ? ''
              : ' ' + JSON.stringify(op.attributes)),
        ),
      )
      .join('\n')
  },
})

describe('markdown', () => {
  test('encode', async () => {
    const parser = await createParser('markdown')
    const contents = await readFile(`${__dirname}/markdown.md`, 'utf-8')
    const [richtext] = transformers.markdown.encode(parser.parse(contents))
    expect(richtext).toMatchSnapshot()
  })

  test('decode', async () => {
    expect(
      transformers.markdown.decode({
        ops: [
          //
          { insert: 'This is ' },
          { insert: 'bold text', attributes: { bold: true } },
          { insert: '.' },
        ],
      }),
    ).toMatchInlineSnapshot(`"This is **bold text**."`)
  })
})
