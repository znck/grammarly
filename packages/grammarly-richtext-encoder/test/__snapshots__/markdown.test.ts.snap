// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown encode 1`] = `
  Heading 1
  \\n {"header":1}
  Heading 2
  \\n {"header":2}
  Heading 5
  \\n {"header":3}
  Heading 4
  \\n {"header":4}
  Heading 5
  \\n {"header":5}
  Heading 6
  \\n {"header":5}
  Heading 1
  \\n {"header":1}
  Heading 2
  \\n {"header":2}
  Inline text can be 
  bold {"bold":true}
  , 
  italic {"italic":true}
  , 
  strikethrough
  , or 
  code {"code":true}
  .
  \\n
  Links: 
  link1 {"link":""}
   
  link2 {"link":"#href"}
  \\n
  Unordered List
  \\n {"list":"bullet","indent":1}
  A
  \\n {"list":"bullet","indent":1}
  A.1
  \\n {"list":"bullet","indent":2}
  A.2
  \\n {"list":"bullet","indent":2}
  B
  \\n {"list":"bullet","indent":1}
  B.1
  \\n {"list":"bullet","indent":2}
  B.2
  \\n {"list":"bullet","indent":2}
  Some 
  inline
   html
  \\n
  First Term
   
  This is the definition of the first term.
  \\n
  Second Term
   
  This is one definition of the second term.
   
  This is another definition of the second term.
  \\n
`;
